fontforgeexe/shapers/harfbuzz.cpp

```diff
@@ -0,0 +1,349 @@
+
+    // Retrieve the current kerning offsets and apply them manually if they
+    // differ from their initial value. The initial value doesn't need
+    // adjustment, since it is hopefully present in the generated font, so that
+    // HarfBuzz takes it into account automatically.
+    std::vector<int> kerning_deltas;
+    for (int i = 0; i + 1 < glyph_count; ++i) {
+        int kerning_offset = context_->get_kern_offset(ots_arr + i);
+        if (kerning_offset == INVALID_KERN_OFFSET) {
+            kerning_offset = 0;
+        }
+
+        // Keep initial kerning offsets when they are first encountered
+        auto key = std::make_pair(glyph_info_arr[i].codepoint,
+                                  glyph_info_arr[i + 1].codepoint);
+        // Insert only if absent
+        initial_kerning_.insert({key, kerning_offset});
+
+        // Compute kerning deltas. Any existing kerning which was not manually
+        // changed by the user should give zero delta.
+        kerning_deltas.push_back(kerning_offset - initial_kerning_[key]);
+    }
+
+    return kerning_deltas;
+}
+
+std::vector<int> HarfBuzzShaper::compute_width_deltas(hb_buffer_t* hb_buffer,
+                                                      SplineChar** glyphs) {
+    unsigned int glyph_count;
+    hb_glyph_info_t* glyph_info_arr =
+        hb_buffer_get_glyph_infos(hb_buffer, &glyph_count);
+
+    // Compute width deltas for glyphs which might have changed.
+    std::vector<int> width_deltas;
+    for (int i = 0; i < glyph_count; ++i) {
+        int16_t width = glyphs[i]->width;
+
+        // Keep initial widths when they are first encountered
+        auto key = glyph_info_arr[i].codepoint;
+        // Insert only if absent
+        initial_width_.insert({key, width});
+
+        // Compute kerning deltas. Any existing kerning which was not manually
+        // changed by the user should give zero delta.
+        width_deltas.push_back(width - initial_width_[key]);
+    }
+
+    return width_deltas;
+}
+
+struct opentype_str* HarfBuzzShaper::apply_features(
+    SplineChar** glyphs, const std::vector<Tag>& feature_list, Tag script,
+    Tag lang, int pixelsize, bool vertical) {
+    std::vector<unichar_t> u_vec;
+    for (size_t len = 0; glyphs[len] != NULL; ++len) {
+        u_vec.push_back(
+            (glyphs[len]->unicodeenc > 0)
+                ? glyphs[len]->unicodeenc
+                : context_->fake_unicode(context_->mv, glyphs[len]));
+    }
+    u_vec.push_back(0);
+
+    char* utf8_str = u2utf8_copy(u_vec.data());
+
+    hb_buffer_t* hb_buffer = hb_buffer_create();
+    hb_buffer_add_utf8(hb_buffer, utf8_str, -1, 0, -1);
+
+    // Set script and language
+    hb_script_t hb_script = hb_script_from_iso15924_tag((uint32_t)script);
+    hb_buffer_set_script(hb_buffer, hb_script);
+    hb_language_t hb_lang = hb_language_from_string((const char*)lang, -1);
+    hb_buffer_set_language(hb_buffer, hb_lang);
+
+    bool rtl = !u_vec.empty() && isrighttoleft(u_vec[0]);
```

Comment:
Checking first character is not enough, for example if text starts with punctuation. HarfBuzz has [`hb_buffer_guess_segment_properties`](https://harfbuzz.github.io/harfbuzz-hb-buffer.html#hb-buffer-guess-segment-properties) which works slightly better here (it checks the first character with strong direction).

Ideally, there should be a UI to set the direction, since it is possible the text is entirely composed of unencoded glyphs which will be given PUA encoding and all PUA codepoints has strong LTR direction.